components/eam/src/physics/cam/ndrop.F90

```diff
@@ -1426,98 +1426,89 @@ subroutine ccncalc(state, pbuf, cs, ccn)
-   real(r8) vaerosol(pcols) ! interstit+activated aerosol volume conc (m3/m3)
-
-   real(r8) amcube(pcols)
-   real(r8) super(psat) ! supersaturation
-   real(r8), allocatable :: amcubecoef(:)
-   real(r8), allocatable :: argfactor(:)
-   real(r8) :: surften       ! surface tension of water w/respect to air (N/m)
-   real(r8) surften_coef
-   real(r8) a(pcols) ! surface tension parameter
-   real(r8) hygro(pcols)  ! aerosol hygroscopicity
-   real(r8) sm(pcols)  ! critical supersaturation at mode radius
-   real(r8) arg(pcols)
-   !     mathematical constants
-   real(r8) twothird,sq2
-   integer l,m,n,i,k
-   real(r8) log,cc
-   real(r8) smcoefcoef,smcoef(pcols)
+   real(r8), pointer :: tair(:,:)     ! air temperature [K]
+   real(r8) naerosol(pcols) ! interstit+activated aerosol number conc [#/m3]
+   real(r8) vaerosol(pcols) ! interstit+activated aerosol volume conc [m3/m3]
+   real(r8) amcube(pcols)  ! [m3]
+   real(r8) super(psat) ! supersaturation [fraction]
+   real(r8) :: amcubecoef(ntot_amode) ! [dimensionless]
+   real(r8) :: argfactor(ntot_amode)  ! [dimensionless]
+   real(r8) surften_coef  ! [m-K]
+   real(r8) aparam(pcols) ! surface tension parameter  [m]
+   real(r8) hygro(pcols)  ! aerosol hygroscopicity [dimensionless]
+   real(r8) sm(pcols)  ! critical supersaturation at mode radius [fraction]
+   real(r8) arg_erf_ccn(pcols) ! [dimensionless] 
+   integer lsat,imode,icol,kk
+   real(r8) smcoefcoef   ! [dimensionless]
+   real(r8) smcoef(pcols)  ! [m^(3/2)]
    integer phase ! phase of aerosol
+   !     mathematical constants
+   real(r8) pct_to_fraction, per_m3_to_per_cm3
+
    !-------------------------------------------------------------------------------
 
+
+   phase=3 ! interstitial+cloudborne
+
    lchnk = state%lchnk
    ncol  = state%ncol
    tair  => state%t
 
-   allocate( &
-      amcubecoef(ntot_amode), &
-      argfactor(ntot_amode)   )
+   pct_to_fraction = 0.01_r8
+   per_m3_to_per_cm3 = 1.e-6_r8 
 
-   super(:)=supersat(:)*0.01_r8
-   sq2=sqrt(2._r8)
-   twothird=2._r8/3._r8
-   surften=0.076_r8
+   super(:)=supersat(:)*pct_to_fraction
    surften_coef=2._r8*mwh2o*surften/(r_universal*rhoh2o)
    smcoefcoef=2._r8/sqrt(27._r8)
 
-   do m=1,ntot_amode
-      amcubecoef(m)=3._r8/(4._r8*pi*exp45logsig(m))
-      argfactor(m)=twothird/(sq2*alogsig(m))
-   end do
+   ccn(:,:,:) = 0._r8
 
-   ccn = 0._r8
-   do k=top_lev,pver
+   do imode=1,ntot_amode
 
-      do i=1,ncol
-         a(i)=surften_coef/tair(i,k)
-         smcoef(i)=smcoefcoef*a(i)*sqrt(a(i))
-      end do
+      amcubecoef(imode)=3._r8/(4._r8*pi*exp45logsig(imode))
+      argfactor(imode)=twothird/(sq2*alogsig(imode))
 
-      do m=1,ntot_amode
-
-         phase=3 ! interstitial+cloudborne
+      do kk=top_lev,pver
 
          call loadaer( &
-            state%q, lchnk, pbuf, 1, ncol, k, &
-            m, nspec_amode(m), cs, phase, naerosol, vaerosol, &
+            state%q, lchnk, pbuf, 1, ncol, kk, &
```

Comment:
We need to push `state%q` and `pbuf` to the interface level. I know it will require some more changes but it will be better if we extract the `state%q` in `state_q` and similarly extract cloud-borne aerosols mmr into `qcldrbn` from pbuf (at the top of this subroutine)and send those instead to `loadaer`. This will require changes to other `loadaer` calls as well. It will also require changes to `loadaer`. Let me know if you have any questions on this.

Reply:
OK, I've pushed the extraction of these fields up to the top level of this file (subroutine dropmixnuc), and changed code in other subroutines accordingly in the latest commit.